File: js/src/model/matching/BestMatch.js
/**
 * Strategy.bestMatch result called by Calculator
 * @author tanzhenlin
 * @date 2022/6/29 11:20
 **/
export default class BestMatch {
    rules;
    items;
    type;
    matches;
    suggestion;

    constructor (rules,items,type) {
        this.rules = rules;
        this.items = items;
        this.type = type;
        this.matches = [];
    }

    /**
     * 添加一条匹配结果
     * @param {Match} match
     */
    addMatch(match) {
        if (match != null) {
            this.matches.push(match);
        }
    }

    /**
     * 被规则命中的所有商品的集合
     * @return {Item []}
     */
    chosen() {
        let chosen = [];
        for (const m of this.matches) {
            for (const t of m.items)
                chosen.push(t);
        }
        return chosen;
    }

    /**
     * 没有被命中的商品的集合
     * @return {Item []}
     */
    left() {
        let chosen = this.chosen();
        return this.items
            .filter(t => !chosen.includes(t));
    }

    /**
     * 所有命中的商品的价格总和
     * @return {number}
     */
    totalPrice() {
        if (this.matches.length === 0) {
            return 0;
        }
        return this.matches
            .map(m => m.totalPrice())
            .reduce((a, b) => a + b, 0);
    }

    /**
     * 总的优惠额
     * @return {number}
     */
    totalDiscount() {
        if (this.matches.length === 0) {
            return 0;
        }
        return this.matches
            .map(m => m.totalDiscount())
            .reduce((a, b) => a + b, 0);
    }
}

File: java/src/main/java/com/github/thinhunan/wonder8/promotion/rule/model/strategy/BestMatch.java
/**
 * Strategy.bestMatch result called by Calculator
 * @author tanzhenlin
 * @date 2022/6/29 11:20
 **/
package com.github.thinhunan.wonder8.promotion.rule.model.strategy;

import com.github.thinhunan.wonder8.promotion.rule.model.Item;
import com.github.thinhunan.wonder8.promotion.rule.model.Rule;
import com.github.thinhunan.wonder8.promotion.rule.model.validate.RuleValidateResult;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class BestMatch {

    MatchType type;

    public MatchType getType() {
        return type;
    }

    public void setType(MatchType type) {
        this.type = type;
    }

    List<Rule> rules;

    public List<Rule> getRules() {
        return rules;
    }

    List<Item> items;

    public List<Item> getItems(){
        return items;
    }

    List<Match> matches;

    public List<Match> getMatches() {
        return matches;
    }

    public void addMatch(Match match){
        if(matches == null){
            matches = new ArrayList<>();
        }
        if(match != null){
            matches.add(match);
        }
    }

    RuleValidateResult suggestion;

    public RuleValidateResult getSuggestion() {
        return suggestion;
    }

    public void setSuggestion(RuleValidateResult suggestion) {
        this.suggestion = suggestion;
    }

    public List<Item> chosen() {
        List<Item> chosen = new ArrayList<>();
        for(Match m : matches){
            chosen.addAll(m.getItems());
        }
        return chosen;
    }

    public List<Item> left() {
        List<Item> chosen = chosen();
        return items.stream()
                .filter(t->!chosen.contains(t))
                .collect(Collectors.toList());
    }

    public BestMatch(List<Rule> rules, List<Item> items, MatchType type) {
        this.rules = rules;
        this.items = items;
        this.type = type;
        this.matches = new ArrayList<>();
    }


    public int totalPrice() {
        if (this.matches == null || this.matches.size() == 0) {
            return 0;
        }
        return matches.stream()
                .map(Match::totalPrice)
                .reduce(0, Integer::sum);
    }

    /**
     * 计算总的打折金额，但请注意只以匹配部分为计算依据，如果需要按所有的商品来计算，可以调用{@link Rule#discount(List)} 来计算
     *
     * @return {int}
     */
    public int totalDiscount() {
        if (this.matches == null || this.matches.size() == 0) {
            return 0;
        }
        return matches.stream()
                .map(Match::totalDiscount)
                .reduce(0, Integer::sum);
    }
}
